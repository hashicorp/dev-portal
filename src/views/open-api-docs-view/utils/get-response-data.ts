/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

// Third party
import { getReasonPhrase } from 'http-status-codes'

// Utils
import { getPropertyDetailPropsFromSchemaObject } from './get-property-detail-props'
// Types
import type { PropertyDetailsGroup } from '../components/operation-details'
import type { PropertyDetailsProps } from '../components/property-details'
import type { OpenAPIV3 } from 'openapi-types'

/**
 * Given OpenAPI responses data,
 * Return response data formatted for display.
 */
export async function getResponseData(
	responses: OpenAPIV3.ResponsesObject,
	slugPrefix: string
): Promise<PropertyDetailsGroup[]> {
	// Set up an object to hold response data
	const responseData: PropertyDetailsGroup[] = []
	// Populate the responseData object using incoming responses
	for (const responseCode of Object.keys(responses)) {
		const value = responses[responseCode]
		// If this value is a reference, skip it
		if ('$ref' in value) {
			continue
		}
		const contentTypeKeys = Object.keys(value.content)

		// If we don't have any content type keys, skip this response
		if (!contentTypeKeys.length) {
			continue
		}

		// If we don't have a definition or schema, skip this response
		const definition = value.content[contentTypeKeys[0]]
		if (!definition || !definition.schema) {
			continue
		}

		// If this schema is a reference, skip this response
		if ('$ref' in definition.schema) {
			continue
		}
		/**
		 * Note: we expect response schemas to always be objects at their root.
		 * We flatten the response properties to avoid showing a redundant object.
		 */
		if (definition.schema.properties) {
			const responseSlug = `${slugPrefix}_${responseCode}`
			const propertyDetails: PropertyDetailsProps[] = []
			const requiredProperties = definition.schema.required || []
			for (const propertyKey of Object.keys(definition.schema.properties)) {
				const data = definition.schema.properties[propertyKey]
				// If this schema is a reference, skip it
				if ('$ref' in data) {
					continue
				}
				const isRequired = requiredProperties.includes(propertyKey)
				propertyDetails.push(
					await getPropertyDetailPropsFromSchemaObject(
						propertyKey,
						data,
						isRequired,
						responseSlug
					)
				)
			}
			// Determine the heading text to show
			const headingText =
				responseCode === 'default'
					? `Default Error Response`
					: `${responseCode} -  ${getReasonPhrase(responseCode)}`
			const headingTheme = responseCode.startsWith('2') ? 'success' : 'critical'

			responseData.push({
				heading: { text: headingText, slug: responseSlug, theme: headingTheme },
				propertyDetails,
			})
		}
	}
	//
	return responseData
}
