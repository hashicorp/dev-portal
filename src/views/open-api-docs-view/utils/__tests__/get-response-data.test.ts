/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import { getResponseData } from '../get-response-data'
import petstore3 from '../__fixtures__/petstore3.openapi'

describe('getResponseData', () => {
	it('parses multiple response definitions for single response code', async () => {
		const request = petstore3.paths['/pet/{petId}'].get
		const groups = await getResponseData(request.responses, 'test-prefix')
		expect(groups).toMatchSnapshot()
	})
})
