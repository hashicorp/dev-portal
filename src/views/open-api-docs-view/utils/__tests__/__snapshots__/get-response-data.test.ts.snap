// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getResponseData parses multiple response definitions for single response code 1`] = `
Array [
  Object {
    "heading": Object {
      "slug": "test-prefix_200",
      "text": "200 -  OK",
      "theme": "success",
    },
    "propertyDetails": Array [
      Object {
        "description": "",
        "isRequired": false,
        "name": "id",
        "nestedProperties": Array [],
        "slug": "test-prefix_200_id",
        "type": "integer",
      },
      Object {
        "description": "",
        "isRequired": true,
        "name": "name",
        "nestedProperties": Array [],
        "slug": "test-prefix_200_name",
        "type": "string",
      },
      Object {
        "description": "",
        "isRequired": false,
        "name": "category",
        "nestedProperties": Array [
          Object {
            "description": "",
            "isRequired": false,
            "name": "id",
            "nestedProperties": Array [],
            "slug": "test-prefix_200_category.id",
            "type": "integer",
          },
          Object {
            "description": "",
            "isRequired": false,
            "name": "name",
            "nestedProperties": Array [],
            "slug": "test-prefix_200_category.name",
            "type": "string",
          },
        ],
        "slug": "test-prefix_200_category",
        "type": "object",
      },
      Object {
        "description": "",
        "isRequired": true,
        "name": "photoUrls",
        "nestedProperties": Array [],
        "slug": "test-prefix_200_photoUrls",
        "type": "string[]",
      },
      Object {
        "description": "",
        "isRequired": false,
        "name": "tags",
        "nestedProperties": Array [
          Object {
            "description": "",
            "isRequired": false,
            "name": "id",
            "nestedProperties": Array [],
            "slug": "test-prefix_200_tags-array.id",
            "type": "integer",
          },
          Object {
            "description": "",
            "isRequired": false,
            "name": "name",
            "nestedProperties": Array [],
            "slug": "test-prefix_200_tags-array.name",
            "type": "string",
          },
        ],
        "slug": "test-prefix_200_tags",
        "type": "object[]",
      },
      Object {
        "description": "<p>pet status in the store</p>",
        "isRequired": false,
        "name": "status",
        "nestedProperties": Array [],
        "slug": "test-prefix_200_status",
        "type": "string",
      },
    ],
  },
]
`;
